### PR TITLE
Add option to set PR to approved or Needs Work

### PR DESCRIPTION
**Pull request by dawidmalina**
Co-authored-by: Dawid Malinowski <dawidmalina@gmail.com>

### CODE REVIEW

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuilds.java
Comment (proski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuilds.java]: As I said on the other PR, I'm not a fan of utility classes. Please move the code from `StashMarkStatus` here. That would be consistent with the code immediately below. If we need to refactor the code, it's better to identify the natural way of doing it, which is easier when similar code resides in one place.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java
Comment (proski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java]: `this` is not needed. The existing code overuses `this`, but it's not a good example to follow IMHO.

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (proski) [inline src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: Please move those entries above `Merge PR if build is successful`. That way, the users would get the idea that the marking would happen before the attempted merge. It's impossible to mark a PR as approved after the merge.

Actually, it's an interesting question whether the merge would be attempted if the PR is not mergeable without the approval from Jenkins. I'm afraid it won't be. It can be implemented separately.

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-approveOnBuildSuccessful.html
Comment (proski) [inline src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-approveOnBuildSuccessful.html]: Please add the final newline.

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-needsWorkOnBuildFailure.html
Comment (proski) [inline src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-needsWorkOnBuildFailure.html]: marked -> mark
"Needs Work" should be quoted or emphasized somehow and possibly capitalized.

File: src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashMarkStatusTest.java
Comment (proski) [inline src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashMarkStatusTest.java]: Please don't use wildcard imports.

File: README.md
Comment (proski) [inline README.md]: Please move those descriptions before `Merge PR if build is successful`, as those events happen before merging. I know, the existing order is not very logical.

**Review by proski — CHANGES_REQUESTED** [review]

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (dawidmalina) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: Good question. On one hand we can block such attempt from plugin side or we can based on bitbucket configuration whether it will allow this action or not.

File: README.md
Comment (dawidmalina) [reply README.md]: done

File: src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashMarkStatusTest.java
Comment (dawidmalina) [reply src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashMarkStatusTest.java]: Test class was removed.

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-needsWorkOnBuildFailure.html
Comment (dawidmalina) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-needsWorkOnBuildFailure.html]: done

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-approveOnBuildSuccessful.html
Comment (dawidmalina) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-approveOnBuildSuccessful.html]: done

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (dawidmalina) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: done

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java
Comment (dawidmalina) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java]: done

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuilds.java
Comment (dawidmalina) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuilds.java]: done

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:454]: if the UnsupportedEncodingException comes from requesting "UTF-8" it's rather unlikely. Please catch it inside the metod and log (not printStackTrace)

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:471]: Please log all exceptions, never print the stack trace to stderr

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-needsWorkOnBuildFailure.html
Comment (jakub-bochenski) [inline src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/help-needsWorkOnBuildFailure.html:2]: > .. "NEEDS WORK" on build failure

or

> "NEEDS WORK" when the build failed

"has been failed" doesn't sound right

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:455]: Please use proper log mesages, "PR-PUT-REQUEST:" is rather cryptic

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:461]: remove the comment please https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:534]: Same here, please use proper sentences

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:524]: more printStackTrace

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:527]: Do not catch Exception, catch specific exceptions

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:481]: This is some bad code. Unfortunately it's a copy paste from other methods for doing http requests. 
Not requesting it for this PR but we definitely need to get rid of this thread spawning.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:499]: I think it actually expects 2xx not strictly 200

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:495]: do not log and throw at the same time
since all other errors are re-thrown I think this one should be too

Comment (jakub-bochenski) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:495]: also the log and exception have different information -- why?

**Comment by jakub-bochenski** [comment]
What I'm still missing is the answer to: why do you need this feature at all. Why not set Stash to only allow merges if there is a successful build?

**Comment by proski** [comment]
> What I'm still missing is the answer to: why do you need this feature at all. Why not set Stash to only allow merges if there is a successful build?

That's a good point. We need a reasonable use case for this patch before it can be merged. The benefits should be weighed against the extra complexity of the configuration.

The standard way for Jenkins to notify Stash of the build results for the purpose of merging is to use special API that is implemented by the stash-notifier plugin. That way, Jenkins is treated as a separate input to the PR, not as a reviewer.

If this feature is used instead of stash-notifier, then reviewers can overrule Jenkins. Jenkins can be removed from the reviewer list by other reviewers. I don't think it's a good idea to have that loophole. If something is broken and Jenkins checks need to be circumvented, it should be done by an admin, not by a reviewer.

**Comment by dawidmalina** [comment]
I don't know about the others intents but in my case the main purpose is to block merge when quality gates are not ok (in this case pull requests validation). This is not something new, take a look https://github.com/mibexsoftware/sonar-bitbucket-plugin as an example. So basically I would love to see exactly the same behaviour with this plugin as I can't use sonar for what I'm working with. If the implementation is not proper please guide me through all your concerns and I will try to tune the code accordingly. We can also take a look how this was done in mentioned plugin.

**Comment by proski** [comment]
@dawidmalina: I've checked that link. I see a screenshot exactly like what I see when stashnotifier-plugin reports failed builds. If you want similar functionality, you should use stashnotifier-plugin to report the build status rather than patch this plugin to provide a poor replacement. Jenkins is not a reviewer, it should talk to Stash through the proper channels.

https://github.com/jenkinsci/stashnotifier-plugin

Speaking of which, we should probably consider merging this plugin with stashnotifier-plugin. That would give us following advantages:

* Unified documentation how to set up Stash pull request support in Jenkins
* Unified configuration - there is a lot of overlap, and there will be more if we add configuration for socket timeouts etc
* Jenkins talking to Stash as one entity, not two
* Less confusion for users
* Developers joining forces
* Possibility to support automated merge when Jenkins requires a successful build (the plugin would notify first, then merge)

I don't see any significant downsides. Larger plugin size maybe, but it's not a big deal.

I wonder what would be the best forum to discuss the potential merge.

As for this PR, I suggest that we close it. Some comments in this PR apply to the existing code. I can take care of the cleanup.

**Comment by jakub-bochenski** [comment]
@proski  agree in general I'm not sure about merging the plugins though.
The notifier plugin seems more fundamental to me. You can use it also for mainline (master, etc.) builds.

Maybe a better idea would be to refactor this plugin to reuse the common parts it has with the notifier plugin (e.g. stash server configuration).
Depending on another plugin rather than merging should also be easier on the communication side.

**Comment by proski** [comment]
The notifier functionality can still be used without the Stash PR trigger if its `isApplicable` method returns `true`. The post-build comment posting should only be applicable with the Stash PR trigger (it is currently `true` unconditionally, which should be fixed).

Depending on stashnotifier-plugin would solve the issue of configuration sharing. However, I hope there is a way to share the configuration without the dependency (e.g. read the configuration by name).

Adding the dependency would not fix the issue of user confusion. Forced installation of stashplugin-notifier would help very little. We would need to improve the documentation for that.

It is so bad that we have a code contribution that duplicates the functionality of stashplugin-notifier. That's a clear sign that our documentation is not good. It's limited to our code, not to the Stash-Jenkins PR build setup as a whole.

**Comment by proski** [comment]
II suggest that we close this PR. It is adding functionality that we cannot and should not support. Instead or adding a poor replacement for stashnotifier-plugin, we should make stashnotifier-plugin more visible in our documentation.

**Comment by jakub-bochenski** [comment]
Agree in general, though I'm a bit busy and trying to find time to reply to your points in https://github.com/jenkinsci/stash-pullrequest-builder-plugin/pull/56#issuecomment-473618219

**Comment by jakub-bochenski** [comment]
> Depending on stashnotifier-plugin would solve the issue of configuration sharing. However, I hope there is a way to share the configuration without the dependency (e.g. read the configuration by name).

I don't follow this at all. Why would we not want to have a dependency here? IMO the ability to seamlessly reuse what other plugins provide is core feature of the plugin system. Plugins depend on other plugins all the time. See e.g. the Git plugin's dependencies

> It is so bad that we have a code contribution that duplicates the functionality of stashplugin-notifier. That's a clear sign that our documentation is not good. It's limited to our code, not to the Stash-Jenkins PR build setup as a whole.

Could you rephrase? I don't understand what point your are trying to make.

**Comment by jakub-bochenski** [comment]
Closing the issue, but please continue with the discussion (unless you think creating a separate issue for this is a better idea)

**Comment by proski** [comment]
> > Depending on stashnotifier-plugin would solve the issue of configuration sharing. However, I hope there is a way to share the configuration without the dependency (e.g. read the configuration by name).
> 
> I don't follow this at all. Why would we not want to have a dependency here? IMO the ability to seamlessly reuse what other plugins provide is core feature of the plugin system. Plugins depend on other plugins all the time. See e.g. the Git plugin's dependencies

Actually yes. Stash Pull Request Builder could depend on Stash Notifier and use the configuration of the later (e.g. the Stash URL).

But it would be a git backwards for a build trigger to use configuration of the post-build action. We would need to document it carefully to minimize confusion.

Another approach would be to implement a separate "Stash service configuration" plugin that would let users configure one or many Stash services in the global Jenkins configuration. Both this plugin and Stash Notifier could use that information.

> > It is so bad that we have a code contribution that duplicates the functionality of stashplugin-notifier. That's a clear sign that our documentation is not good. It's limited to our code, not to the Stash-Jenkins PR build setup as a whole.
> 
> Could you rephrase? I don't understand what point your are trying to make.

My point is:

* We ignore Stash Notifier in the documentation because it's a separate plugin
* Users set up this plugin and not Stash Notifier
* Users want to have PR approval/disapproval base on the build status
* Somebody writes a long piece of code to make this plugin approve/disapprove PRs
* All that work is wasted because Stash Notifier does it the right way
* We should not ignore Stash Notifier in the documentation to avoid such situations

**Comment by jakub-bochenski** [comment]
EDIT: let's continue in #52

**Pull request closed without merging**
